server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -102,14 +127,122 @@ public void writeFileChunk(
+
+        // Get list of files to copy from this checkpoint.
+        source.getCheckpointMetadata(getId(), checkpoint, checkpointInfoListener);
+
+        checkpointInfoListener.whenComplete(checkpointInfo -> getFiles(checkpointInfo, getFilesListener), listener::onFailure);
+        getFilesListener.whenComplete(
+            response -> finalizeReplication(checkpointInfoListener.result(), finalizeListener),
+            listener::onFailure
+        );
+        finalizeListener.whenComplete(r -> listener.onResponse(null), listener::onFailure);
+    }
+
+    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener)
+        throws IOException {
+        final Store.MetadataSnapshot snapshot = checkpointInfo.getSnapshot();
+        Store.MetadataSnapshot localMetadata = getMetadataSnapshot();
+        final Store.RecoveryDiff diff = snapshot.recoveryDiff(localMetadata);
+        logger.debug("Recovery diff {}", diff);
+        final List<StoreFileMetadata> filesToFetch = Stream.concat(diff.missing.stream(), diff.different.stream())
+            .collect(Collectors.toList());
+
+        Set<String> storeFiles = new HashSet<>(Arrays.asList(store.directory().listAll()));
+        final Set<StoreFileMetadata> pendingDeleteFiles = checkpointInfo.getPendingDeleteFiles()
+            .stream()
+            .filter(f -> storeFiles.contains(f.name()) == false)
+            .collect(Collectors.toSet());
+
+        filesToFetch.addAll(pendingDeleteFiles);
+
+        for (StoreFileMetadata file : filesToFetch) {
+            state.getIndex().addFileDetail(file.name(), file.length(), false);
+        }
+        if (filesToFetch.isEmpty()) {
+            getFilesListener.onResponse(new GetSegmentFilesResponse(filesToFetch));
+        }
+        source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), filesToFetch, store, getFilesListener);
+    }
+
+    private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse, ActionListener<Void> listener) {
+        ActionListener.completeWith(listener, () -> {
+            // first, we go and move files that were created with the recovery id suffix to
+            // the actual names, its ok if we have a corrupted index here, since we have replicas
+            // to recover from in case of a full cluster shutdown just when this code executes...
+            multiFileWriter.renameAllTempFiles();
+            final Store store = store();
+            store.incRef();
+            try {
+                // Deserialize the new SegmentInfos object sent from the primary.
+                final ReplicationCheckpoint responseCheckpoint = checkpointInfoResponse.getCheckpoint();
+                SegmentInfos infos = SegmentInfos.readCommit(
+                    store.directory(),
+                    toIndexInput(checkpointInfoResponse.getInfosBytes()),
+                    responseCheckpoint.getSegmentsGen()
+                );
```

Comment:
we have a store reference here already, and this method is intended to perform cleanup & finalize the replication. I would consider moving the store.cleanupAndVerify call here instead of the engine. Then we don't need the snapshot in the shard.finalize method.